cljs: Replace clojure.spec leftovers with Malli

alexander-yakushev opened the pull request:
There are currently only two CLJS namespaces that reference clojure.spec. Getting rid of clojure.spec saves 81KB (18KB gzipped) in the final bundle.

alexander-yakushev committed:
cljs: Replace clojure.spec leftovers with Malli

alexander-yakushev commented on src/metabase/models/visualization_settings.cljc:
```diff
@@ -23,144 +23,139 @@
+   [::column-settings {:optional true} map?]
+   [::click-behavior {:optional true} click-behavior-schema]])
+
+;; (s/def ::link-target-id (s/or :int int? :fully-qualified-name string?))
+;; (s/def ::link-template string?)
+;; (s/def ::link-text-template string?)
+
+;; (s/def ::column-title string?)
+;; (s/def ::date-style #{"M/D/YYYY" "D/M/YYYY" "YYYY/M/D" "MMMM D, YYYY" "D MMMM, YYYY" "dddd, MMMM D, YYYY"})
+;; (s/def ::date-abbreviate boolean?)
+;; (s/def ::date-separator #{"/" "-" "."})
+;; (s/def ::time-style #{"HH:mm" "h:mm A" "h A"})
+;; (s/def ::time-enabled #{nil "minutes" "seconds" "milliseconds"})
+;; (s/def ::decimals pos-int?)
+;; (s/def ::number-separators #(or nil? (and string? (= 2 (count %)))))
+;; (s/def ::number-style #{"decimal" "percent" "scientific" "currency"})
+;; (s/def ::prefix string?)
+;; (s/def ::suffix string?)
+;; (s/def ::view-as string?)
+;; (s/def ::link-text string?)
+
+;; (s/def ::param-mapping-id string?)
+
+;; (s/def ::param-ref-type #{"column" "dimension" "variable" "parameter"})
+;; (s/def ::param-ref-id string?)
+;; (s/def ::param-ref-name string?)
```
I'm not sure what to do with these. These specs are not parts of other specs and are not references by any of the fdef's. I've commented them because they look important (they describe the data that we expect) but it would be better if they validated something too.

camsaul replied:
Some of these like `::link-text` were being used by the `::click-behavior` spec but since you removed it probably ok to remove these entirely. As far as I can tell `::number-style` was never used at all https://github.com/metabase/metabase/pull/15858/files#diff-b1e5711638b653b2a67d36d1b997a409c2625141fca2307ff3fe46c017249de9R381

camsaul commented on src/metabase/util/log/capture.cljc:
```diff
@@ -140,10 +143,6 @@
      `(do ~@body)
      bindings)))
 
-(s/fdef with-log-messages-for-level
-  :args ::with-log-messages-for-level-args
-  :ret  any?)
```
I still think `fdef` on a macro is still useful because the validation on macroexpansion is actually built in to Clojure. Since macroexpansion only happens on the JVM anyway can't we just make `::with-log-messages-for-level-args` and friends `#?:clj` so we can keep this the way it is but still remove the Cljs need for spec?

camsaul reviewed (commented):
Generally 👍 on this with two caveats

- I would prefer to leave the `fdef`s on the `defmacro`s, but that shouldn't affect your ability to make remove the Cljs dep on spec
- Once https://github.com/clj-kondo/clj-kondo/issues/2521 lands then we should add `clojure.core.spec.alpha` to the Kondo discouraged namespaces for cljs... can you open an issue so we don't forget to do this?

alexander-yakushev commented:
Thanks for the review!

I've addressed your comments.
- Created https://github.com/metabase/metabase/issues/58538
- Reverted spec in `capture.cljc` making it clj-only.
- Removed those specs that I previously commented out.
